monix-iterant/shared/src/main/scala/monix/iterant/Iterant.scala

```diff
@@ -0,0 +1,623 @@
+/*
+ * Copyright (c) 2014-2017 by its authors. Some rights reserved.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.iterant
+
+import monix.eval.{Coeval, Task}
+import monix.types._
+import monix.types.syntax._
+import scala.collection.immutable.LinearSeq
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+/** The `Iterant` is a type that describes lazy, possibly asynchronous
+  * streaming of elements.
+  *
+  * It is similar somewhat in spirit to Scala's own
+  * `collection.immutable.Stream` and with Java's `Iterable`, except
+  * that it is more composable and more flexible due to evaluation being
+  * controlled by an `F[_]` monadic type that you have to supply
+  * (like [[monix.eval.Task Task]] or [[monix.eval.Coeval Coeval]])
+  * which will control the evaluation. In other words, this `Iterant`
+  * type is capable of strict or lazy, synchronous or asynchronous
+  * evaluation.
+  *
+  * Consumption of a `Iterant` happens typically in a loop where
+  * the current step represents either a signal that the stream
+  * is over, or a (head, rest) pair, very similar in spirit to
+  * Scala's standard `List` or `Iterant`.
```

Comment:
"standard `List` or `Iterant`." I presume you meant `Iterator` here, not `Iterant`.